### DISCUSSION TITLE
Value Tuple Field XML docs via Alias

### DISCUSSION THREAD

**Discussion by ZacharyPatten** [dotnet/csharplang#4044] [Language Ideas]
> I tried to comment on #145, but GitHub is blocking me from doing so (I assume it is a bug with GitHub's `Discussions` feature)...

In terms of documenting the fields of value tuples, it seems like the best place to do that might be on alias's. That way it works for more use cases than just return values on methods. For example it would also work when you used the alias inside a `List<MYALIAS>`.

Here is an example of what it could look like. I used a standard `TryParse` method as an example, but this would be nice for any method intended to have multiple return values (or any time the alias was used).
```cs
using System;
/// <field name="Success">I'm the xml docs for Success.</param>
/// <field name="Value">I'm the xml docs for Value.</param>
using Result = (bool Success, int Value);

class Program
{
	static void Main(string[] args)
	{
		var result = TryParse("5");
		Console.WriteLine(result.Success);
		Console.WriteLine(result.Value);
	}

	public static Result TryParse(string input) =>
		(int.TryParse(input, out int value), value);
}
```

Alias's doesn't support Value Tuple syntax at the moment, but if/when that is added, maybe a feature like this could be included.

Another issue is that alias's don't support generics at the moment, but I'm under the impression that is a potential future feature for C#. Thus the following would not be possible until generic alias's were added:
```cs
using System;
/// <field name="Success">I'm the xml docs for Success.</param>
/// <field name="Value">I'm the xml docs for Value.</param>
using Result<T> = (bool Success, T Value);

class Program
{
	static void Main(string[] args)
	{
		var result = TryParse<double>("5");
		Console.WriteLine(result.Success);
		Console.WriteLine(result.Value);
	}

	public static Result<T> TryParse<T>(string input) =>
		/* code ommitted to keep example short */;
}
```

So this topic should likely be on hold until alias's support those two topics...

- value tuple syntax
- generics

...but when they do, XML docs for value tuple fields might be a nice next step.

**Comment by YairHalberstadt** [comment]
At the moment aliases are single file only, so benefit of xml docs on them seems limited...

If we ever get public aliases, I think this would be much more valuable.

**Reply by ZacharyPatten** [reply]
Although I am in favor of cross-file alias's :) , I was imagining it not be necessary for this feature. Just because the XML would be on an alias doesn't mean IDEs couldn't still use the XML in the calling code that doesn't use the alias. The alias would just be a way to set the XML if XML was desired. So if the calling code deconstructed the value tuple...
```cs
var (success, value) = TryParse("5");
```
...I see no reason why IDEs couldn't still have hover-over docs for each local of the deconstruction. Or even if `var` was used (no deconstruction) the XML docs could also be retained.

**Reply by ZacharyPatten** [reply]
@YairHalberstadt  We now have...
```cs
global using Result = System.ValueTuple<bool, int>;
```
...in .NET 6. Now we just need https://github.com/dotnet/csharplang/issues/4284 (and https://github.com/dotnet/csharplang/issues/1239 would be nice too). :P

See you next year. :D

**Comment by dstarkowski** [comment]
I fail to understand why not simply create a `struct Result<T>` in this example. It's reusable. It's generic. It supports docs. If you need to define and document an alias, defining and documenting a struct doesn't seem like much more of a hassle. 

Is tuple even considered good choice for public apis?

**Reply by ZacharyPatten** [reply]
People don't always want to bloat their libraries with a bunch of types purely to wrap multiple return values. Using value tuple can result in more readable code because you can decompose the value tuple...
```cs
var (success, value) = TryParse("5");
```
...so it is very easy to read that the method returns multiple values.

I could use your argument to argue against `out` parameters being in the language too, but those are a feature and are still being used today.

Personally I see nothing wrong with using value tuple for return types in C# even in public APIs. See #145 if you did not already.

**Reply by chrisoverzero** [reply]
>Using value tuple can result in more readable code because you can decompose the value tuple...

You can deconstruct any type you like with the `Deconstruct` [method][] or [extension method][].

[method]: https://docs.microsoft.com/en-us/dotnet/csharp/deconstruct#deconstructing-user-defined-types
[extension method]: https://docs.microsoft.com/en-us/dotnet/csharp/deconstruct#deconstructing-a-user-defined-type-with-an-extension-method

**Reply by ZacharyPatten** [reply]
@chrisoverzero 
Yes but that still means that you need to have:
- a custom type for every method with unique returns
- a destruct method on all the types
- overrides to work like ValueTuple if desired (at least ToString and maybe GetHashCode)
- even more XML docs (now you have new types and new methods that should be documented along with the fields)

ValueTuple syntax let's people ignore all the extra necessities as to not bloat a code base with needless definitions. Value Tuples just don't support XML docs on named fields.

~~Also (most importantly), once you destruct a custom type you lose the XML on those fields.~~ _**Edit**: This point is bad because it could be added to custom type deconstruction too._

**Reply by chrisoverzero** [reply]
>Yes but that still means that you need to have:

I don't find those to be burdensome.
